lib/grant-types/authorization-code-grant-type.ts

```diff
@@ -0,0 +1,202 @@
+        'Invalid argument: model does not implement `revokeAuthorizationCode()`',
+      );
+    }
+
+    if (!options.model.saveToken) {
+      throw new InvalidArgumentError(
+        'Invalid argument: model does not implement `saveToken()`',
+      );
+    }
+  }
+
+  /**
+   * Handle authorization code grant.
+   *
+   * @see https://tools.ietf.org/html/rfc6749#section-4.1.3
+   */
+
+  async handle(request: Request, client: Client) {
+    if (!request) {
+      throw new InvalidArgumentError('Missing parameter: `request`');
+    }
+
+    if (!client) {
+      throw new InvalidArgumentError('Missing parameter: `client`');
+    }
+    const code = await this.getAuthorizationCode(request, client);
+    this.validateRedirectUri(request, code);
+    await this.revokeAuthorizationCode(code);
+
+    return this.saveToken(
+      code.user,
+      client,
+      code.authorizationCode,
+      code.scope,
+    );
+  }
+
+  /**
+   * Get the authorization code.
+   */
+
+  async getAuthorizationCode(request: Request, client: Client) {
+    if (!request.body.code) {
+      throw new InvalidRequestError('Missing parameter: `code`');
+    }
+
+    if (!is.vschar(request.body.code)) {
+      throw new InvalidRequestError('Invalid parameter: `code`');
+    }
+
+    const code = await this.model.getAuthorizationCode(request.body.code);
+    if (!code) {
+      throw new InvalidGrantError(
+        'Invalid grant: authorization code is invalid',
+      );
+    }
+
+    if (!code.client) {
+      throw new ServerError(
+        'Server error: `getAuthorizationCode()` did not return a `client` object',
+      );
+    }
+
+    if (!code.user) {
+      throw new ServerError(
+        'Server error: `getAuthorizationCode()` did not return a `user` object',
+      );
+    }
+
+    if (code.client.id !== client.id) {
+      throw new InvalidGrantError(
+        'Invalid grant: authorization code is invalid',
+      );
+    }
+
+    if (!(code.expiresAt instanceof Date)) {
+      throw new ServerError(
+        'Server error: `expiresAt` must be a Date instance',
+      );
+    }
+
+    if (code.expiresAt.getTime() < Date.now()) {
+      throw new InvalidGrantError(
+        'Invalid grant: authorization code has expired',
+      );
+    }
+
+    if (code.redirectUri && !is.uri(code.redirectUri)) {
+      throw new InvalidGrantError(
+        'Invalid grant: `redirect_uri` is not a valid URI',
+      );
+    }
+
+    return code;
+  }
+
+  /**
+   * Validate the redirect URI.
+   *
+   * "The authorization server MUST ensure that the redirect_uri parameter is
+   * present if the redirect_uri parameter was included in the initial
+   * authorization request as described in Section 4.1.1, and if included
+   * ensure that their values are identical."
+   *
+   * @see https://tools.ietf.org/html/rfc6749#section-4.1.3
+   */
+
+  validateRedirectUri(request: Request, code: AuthorizationCode) {
+    if (!code.redirectUri) {
+      return;
```

Comment:
Should this return (void) or an actual (`it === false`), i.e. a real bool?

Reply:
if code.redirectUri is undefined then a strict test to false would be not return null but the method would go on processing with the following code and would throw the InvalidRequestError in line 148, even though it would be valid request, as there was no redirectUri in the initial authorization (as described in the comment in line 126 - 129). 

This method returns null / void, as the calling method is handling the thrown errors and not the possible return values of this method.